File: apps/macos/Packages/DailyDoListModels/Sources/DailyDoListModels/Events.swift
import Foundation

public struct VaultChangeOrigin: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let external: Self = "external"
  public static let client: Self = "client"
  public static let agent: Self = "agent"
  public static let sync: Self = "sync"
}

public enum VaultChangeKind: String, Codable, Hashable, Sendable {
  case created, modified, deleted
}

public struct VaultChange: Codable, Hashable, Sendable {
  public var path: String
  public var kind: VaultChangeKind
  public var version: String?

  public init(path: String, kind: VaultChangeKind, version: String? = nil) {
    self.path = path
    self.kind = kind
    self.version = version
  }
}

/// Machine-readable reason of a server `error` event.
public struct WsErrorCode: WireEnum {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }
  public static let invalidJSON: Self = "invalid_json"
  public static let invalidMessage: Self = "invalid_message"
  public static let binaryUnsupported: Self = "binary_unsupported"
  public static let tooManySubscriptions: Self = "too_many_subscriptions"
  public static let subscribeFailed: Self = "subscribe_failed"
  public static let incompatibleApiVersion: Self = "incompatible_api_version"
}

// MARK: - Server events

public struct HelloEvent: Codable, Hashable, Sendable {
  public var serverVersion: String
  public var apiVersion: Int
  public init(serverVersion: String, apiVersion: Int) {
    self.serverVersion = serverVersion
    self.apiVersion = apiVersion
  }
}

public struct VaultChangedEvent: Codable, Hashable, Sendable {
  public var changes: [VaultChange]
  public var origin: VaultChangeOrigin
  /// Client that caused the change (so it can ignore its own echo).
  public var clientId: String?
  public init(changes: [VaultChange], origin: VaultChangeOrigin, clientId: String? = nil) {
    self.changes = changes
    self.origin = origin
    self.clientId = clientId
  }
}

public struct TaskRecordsEvent: Codable, Hashable, Sendable {
  public var notePath: String
  public var records: [TaskAgentRecord]
  public init(notePath: String, records: [TaskAgentRecord]) {
    self.notePath = notePath
    self.records = records
  }
}

public struct ThreadMessageEvent: Codable, Hashable, Sendable {
  public var threadId: String
  public var message: ThreadMessage
  public init(threadId: String, message: ThreadMessage) {
    self.threadId = threadId
    self.message = message
  }
}

public struct ThreadDeltaEvent: Codable, Hashable, Sendable {
  public var threadId: String
  public var messageId: String
  public var delta: String
  public init(threadId: String, messageId: String, delta: String) {
    self.threadId = threadId
    self.messageId = messageId
    self.delta = delta
  }
}

public struct ServerErrorEvent: Codable, Hashable, Sendable {
  public var message: String
  public var code: WsErrorCode?
  public init(message: String, code: WsErrorCode? = nil) {
    self.message = message
    self.code = code
  }
}

/// Everything the daemon pushes on `/ws`, discriminated by `type`.
public enum ServerEvent: Hashable, Sendable {
  /// First event on every connection.
  case hello(HelloEvent)
  case vaultChanged(VaultChangedEvent)
  case taskRecords(TaskRecordsEvent)
  case taskRecord(TaskAgentRecord)
  case threadUpsert(ThreadSummary)
  case threadMessage(ThreadMessageEvent)
  case threadDelta(ThreadDeltaEvent)
  case approvalUpsert(ApprovalRequest)
  case agentStatus(AgentStatusResponse)
  case surfaceFrame(SurfaceFrame)
  case settingsChanged(AppSettings)
  /// Every routine, whenever one changed.
  case routinesChanged([Routine])
  /// A run finished and its routine's `notify` says to tell the user.
  case routineNotification(RoutineNotification)
  case error(ServerErrorEvent)
  /// A type this client doesn't know (from a newer daemon): ignore it.
  case unknown(type: String, raw: JSONValue)

  public var type: String {
    switch self {
    case .hello: "hello"
    case .vaultChanged: "vault.changed"
    case .taskRecords: "task.records"
    case .taskRecord: "task.record"
    case .threadUpsert: "thread.upsert"
    case .threadMessage: "thread.message"
    case .threadDelta: "thread.delta"
    case .approvalUpsert: "approval.upsert"
    case .agentStatus: "agent.status"
    case .surfaceFrame: "surface.frame"
    case .settingsChanged: "settings.changed"
    case .routinesChanged: "routines.changed"
    case .routineNotification: "routine.notification"
    case .error: "error"
    case .unknown(let type, _): type
    }
  }
}

extension ServerEvent: Codable {
  private enum Keys: String, CodingKey {
    case type, record, thread, approval, status, settings, routines, notification
  }

  public init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: Keys.self)
    let type = try c.decode(String.self, forKey: .type)
    switch type {
    case "hello": self = .hello(try HelloEvent(from: decoder))
    case "vault.changed": self = .vaultChanged(try VaultChangedEvent(from: decoder))
    case "task.records": self = .taskRecords(try TaskRecordsEvent(from: decoder))
    case "task.record": self = .taskRecord(try c.decode(TaskAgentRecord.self, forKey: .record))
    case "thread.upsert": self = .threadUpsert(try c.decode(ThreadSummary.self, forKey: .thread))
    case "thread.message": self = .threadMessage(try ThreadMessageEvent(from: decoder))
    case "thread.delta": self = .threadDelta(try ThreadDeltaEvent(from: decoder))
    case "approval.upsert":
      self = .approvalUpsert(try c.decode(ApprovalRequest.self, forKey: .approval))
    case "agent.status":
      self = .agentStatus(try c.decode(AgentStatusResponse.self, forKey: .status))
    case "surface.frame": self = .surfaceFrame(try SurfaceFrame(from: decoder))
    case "settings.changed":
      self = .settingsChanged(try c.decode(AppSettings.self, forKey: .settings))
    case "routines.changed":
      self = .routinesChanged(try c.decode([Routine].self, forKey: .routines))
    case "routine.notification":
      self = .routineNotification(
        try c.decode(RoutineNotification.self, forKey: .notification))
    case "error": self = .error(try ServerErrorEvent(from: decoder))
    default: self = .unknown(type: type, raw: try JSONValue(from: decoder))
    }
  }

  public func encode(to encoder: Encoder) throws {
    switch self {
    case .hello(let e): try encodeTagged(e, kind: "hello", key: "type", to: encoder)
    case .vaultChanged(let e): try encodeTagged(e, kind: "vault.changed", key: "type", to: encoder)
    case .taskRecords(let e): try encodeTagged(e, kind: "task.records", key: "type", to: encoder)
    case .taskRecord(let record):
      try encodeWrapped(record, key: .record, type: "task.record", to: encoder)
    case .threadUpsert(let thread):
      try encodeWrapped(thread, key: .thread, type: "thread.upsert", to: encoder)
    case .threadMessage(let e):
      try encodeTagged(e, kind: "thread.message", key: "type", to: encoder)
    case .threadDelta(let e): try encodeTagged(e, kind: "thread.delta", key: "type", to: encoder)
    case .approvalUpsert(let approval):
      try encodeWrapped(approval, key: .approval, type: "approval.upsert", to: encoder)
    case .agentStatus(let status):
      try encodeWrapped(status, key: .status, type: "agent.status", to: encoder)
    case .surfaceFrame(let frame):
      try encodeTagged(frame, kind: "surface.frame", key: "type", to: encoder)
    case .settingsChanged(let settings):
      try encodeWrapped(settings, key: .settings, type: "settings.changed", to: encoder)
    case .routinesChanged(let routines):
      try encodeWrapped(routines, key: .routines, type: "routines.changed", to: encoder)
    case .routineNotification(let notification):
      try encodeWrapped(
        notification, key: .notification, type: "routine.notification", to: encoder)
    case .error(let e): try encodeTagged(e, kind: "error", key: "type", to: encoder)
    case .unknown(_, let raw): try raw.encode(to: encoder)
    }
  }

  private func encodeWrapped<T: Encodable>(_ value: T, key: Keys, type: String, to encoder: Encoder)
    throws
  {
    var c = encoder.container(keyedBy: Keys.self)
    try c.encode(type, forKey: .type)
    try c.encode(value, forKey: key)
  }
}

// MARK: - Client events

/// Signals a client sends on `/ws`.
public enum ClientEvent: Hashable, Sendable {
  case hello(
    clientId: String, apiVersion: Int = DaemonProtocol.apiVersion, clientVersion: String? = nil)
  case ping
  case surfaceSubscribe(threadId: String, surface: SurfaceKind)
  case surfaceUnsubscribe(threadId: String, surface: SurfaceKind)
  case threadRead(threadId: String)
  /// Where the user is typing, so the orchestrator never jumps on a half-written task.
  case editorActivity(notePath: String, line: Int)

  public var type: String {
    switch self {
    case .hello: "hello"
    case .ping: "ping"
    case .surfaceSubscribe: "surface.subscribe"
    case .surfaceUnsubscribe: "surface.unsubscribe"
    case .threadRead: "thread.read"
    case .editorActivity: "editor.activity"
    }
  }
}

extension ClientEvent: Codable {
  private enum Keys: String, CodingKey {
    case type, clientId, apiVersion, clientVersion, threadId, surface, notePath, line
  }

  public init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: Keys.self)
    let type = try c.decode(String.self, forKey: .type)
    switch type {
    case "hello":
      self = .hello(
        clientId: try c.decode(String.self, forKey: .clientId),
        apiVersion: try c.decodeIfPresent(Int.self, forKey: .apiVersion) ?? 1,
        clientVersion: try c.decodeIfPresent(String.self, forKey: .clientVersion))
    case "ping": self = .ping
    case "surface.subscribe":
      self = .surfaceSubscribe(
        threadId: try c.decode(String.self, forKey: .threadId),
        surface: try c.decode(SurfaceKind.self, forKey: .surface))
    case "surface.unsubscribe":
      self = .surfaceUnsubscribe(
        threadId: try c.decode(String.self, forKey: .threadId),
        surface: try c.decode(SurfaceKind.self, forKey: .surface))
    case "thread.read": self = .threadRead(threadId: try c.decode(String.self, forKey: .threadId))
    case "editor.activity":
      self = .editorActivity(
        notePath: try c.decode(String.self, forKey: .notePath),
        line: try c.decode(Int.self, forKey: .line))
    default:
      throw DecodingError.dataCorruptedError(
        forKey: .type, in: c, debugDescription: "Unknown client event type \(type)")
    }
  }

  public func encode(to encoder: Encoder) throws {
    var c = encoder.container(keyedBy: Keys.self)
    try c.encode(type, forKey: .type)
    switch self {
    case .hello(let clientId, let apiVersion, let clientVersion):
      try c.encode(clientId, forKey: .clientId)
      try c.encode(apiVersion, forKey: .apiVersion)
      try c.encodeIfPresent(clientVersion, forKey: .clientVersion)
    case .ping:
      break
    case .surfaceSubscribe(let threadId, let surface),
      .surfaceUnsubscribe(let threadId, let surface):
      try c.encode(threadId, forKey: .threadId)
      try c.encode(surface, forKey: .surface)
    case .threadRead(let threadId):
      try c.encode(threadId, forKey: .threadId)
    case .editorActivity(let notePath, let line):
      try c.encode(notePath, forKey: .notePath)
      try c.encode(line, forKey: .line)
    }
  }
}

// MARK: - JSON coding helpers

extension JSONDecoder {
  /// Decoder configured for daemon payloads.
  public static var daemon: JSONDecoder { JSONDecoder() }
}

extension JSONEncoder {
  /// Encoder configured for daemon payloads (stable key order for tests and logs).
  public static var daemon: JSONEncoder {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
    return encoder
  }
}

File: packages/contract/src/wire/events.ts
/**
 * WebSocket messages. Server events tolerate unknown keys (a newer daemon may add fields), and
 * clients ignore event types they don't know. Client events are strict.
 */
import { z } from "zod";
import {
  ApprovalRequestSchema,
  RoutineNotificationSchema,
  RoutineSchema,
  SurfaceKindSchema,
  surfaceFrameShape,
  TaskAgentRecordSchema,
  ThreadMessageSchema,
  ThreadSummarySchema,
} from "./domain";
import { ObsidianImportJobSchema } from "./imports";
import {
  ClientIdSchema,
  ContentVersionSchema,
  CountSchema,
  IdSchema,
  RequestPathSchema,
  RuntimeIdSchema,
  VaultPathSchema,
} from "./primitives";
import { named } from "./registry";
import { AgentStatusResponseSchema } from "./rest";
import { AppSettingsSchema } from "./settings";

export const VaultChangeOriginSchema = named(
  "VaultChangeOrigin",
  "Who changed the vault: another program, a client (see `clientId`), the agent, or sync.",
  z.enum(["external", "client", "agent", "sync"]),
);

export const VaultChangeSchema = named(
  "VaultChange",
  "One changed path.",
  z.looseObject({
    path: VaultPathSchema,
    kind: z.enum(["created", "modified", "deleted"]),
    version: ContentVersionSchema.optional(),
  }),
);

export const WsErrorCodeSchema = named(
  "WsErrorCode",
  "Machine-readable reason of a server `error` event.",
  z.enum([
    "invalid_json",
    "invalid_message",
    "binary_unsupported",
    "too_many_subscriptions",
    "subscribe_failed",
    "incompatible_api_version",
  ]),
);

// ── Server → client ──────────────────────────────────────────────────────

export const ServerHelloEventSchema = named(
  "ServerHelloEvent",
  "First event on every connection.",
  z.looseObject({
    type: z.literal("hello"),
    serverVersion: z.string().min(1).max(100),
    apiVersion: z.int().min(1),
  }),
);

export const VaultChangedEventSchema = named(
  "VaultChangedEvent",
  "Vault paths changed (coalesced). Clients skip batches carrying their own `clientId`.",
  z.looseObject({
    type: z.literal("vault.changed"),
    changes: z.array(VaultChangeSchema),
    origin: VaultChangeOriginSchema,
    clientId: ClientIdSchema.optional(),
  }),
);

export const TaskRecordsEventSchema = named(
  "TaskRecordsEvent",
  "Snapshot of a note's task records.",
  z.looseObject({
    type: z.literal("task.records"),
    notePath: VaultPathSchema,
    records: z.array(TaskAgentRecordSchema),
  }),
);

export const TaskRecordEventSchema = named(
  "TaskRecordEvent",
  "One task record changed.",
  z.looseObject({ type: z.literal("task.record"), record: TaskAgentRecordSchema }),
);

export const ThreadUpsertEventSchema = named(
  "ThreadUpsertEvent",
  "A thread was created or its summary changed.",
  z.looseObject({ type: z.literal("thread.upsert"), thread: ThreadSummarySchema }),
);

export const ThreadMessageEventSchema = named(
  "ThreadMessageEvent",
  "A message was added or replaced (same `id`) in a thread.",
  z.looseObject({
    type: z.literal("thread.message"),
    threadId: RuntimeIdSchema,
    message: ThreadMessageSchema,
  }),
);

export const ThreadDeltaEventSchema = named(
  "ThreadDeltaEvent",
  "Streaming text appended to a `streaming` text message (droppable; the final message repairs gaps).",
  z.looseObject({
    type: z.literal("thread.delta"),
    threadId: RuntimeIdSchema,
    messageId: IdSchema,
    delta: z.string(),
  }),
);

export const ApprovalUpsertEventSchema = named(
  "ApprovalUpsertEvent",
  "An approval request was created or decided.",
  z.looseObject({ type: z.literal("approval.upsert"), approval: ApprovalRequestSchema }),
);

export const AgentStatusEventSchema = named(
  "AgentStatusEvent",
  "The agent status changed.",
  z.looseObject({ type: z.literal("agent.status"), status: AgentStatusResponseSchema }),
);

export const SurfaceFrameEventSchema = named(
  "SurfaceFrameEvent",
  "A live surface frame; only sent to clients subscribed to that thread's surface (droppable).",
  z.looseObject({ type: z.literal("surface.frame"), ...surfaceFrameShape }),
);

export const SettingsChangedEventSchema = named(
  "SettingsChangedEvent",
  "The effective settings changed.",
  z.looseObject({ type: z.literal("settings.changed"), settings: AppSettingsSchema }),
);

export const RoutinesChangedEventSchema = named(
  "RoutinesChangedEvent",
  "Every routine, whenever one changed (its file, its schedule, its last run).",
  z.looseObject({ type: z.literal("routines.changed"), routines: z.array(RoutineSchema) }),
);

export const RoutineNotificationEventSchema = named(
  "RoutineNotificationEvent",
  "A routine's run finished and its `notify` says to tell the user (clients show a notification).",
  z.looseObject({
    type: z.literal("routine.notification"),
    notification: RoutineNotificationSchema,
  }),
);

export const ImportProgressEventSchema = named(
  "ImportProgressEvent",
  "An import or update from Obsidian progressed (at most every 200 ms), changed phase, or ended (`job.state`).",
  z.looseObject({ type: z.literal("import.progress"), job: ObsidianImportJobSchema }),
);

export const ServerErrorEventSchema = named(
  "ServerErrorEvent",
  "Something the client sent was rejected (or the connection is about to close).",
  z.looseObject({
    type: z.literal("error"),
    message: z.string(),
    code: WsErrorCodeSchema.optional(),
  }),
);

export const ServerEventSchema = named(
  "ServerEvent",
  "Every server → client WebSocket message, discriminated by `type`.",
  z.discriminatedUnion("type", [
    ServerHelloEventSchema,
    VaultChangedEventSchema,
    TaskRecordsEventSchema,
    TaskRecordEventSchema,
    ThreadUpsertEventSchema,
    ThreadMessageEventSchema,
    ThreadDeltaEventSchema,
    ApprovalUpsertEventSchema,
    AgentStatusEventSchema,
    SurfaceFrameEventSchema,
    SettingsChangedEventSchema,
    RoutinesChangedEventSchema,
    RoutineNotificationEventSchema,
    ImportProgressEventSchema,
    ServerErrorEventSchema,
  ]),
);

// ── Client → server ──────────────────────────────────────────────────────

export const ClientHelloEventSchema = named(
  "ClientHelloEvent",
  "First message a client sends: its id and the API version it speaks.",
  z.strictObject({
    type: z.literal("hello"),
    clientId: ClientIdSchema,
    apiVersion: z
      .int()
      .min(1)
      .optional()
      .describe("API_VERSION the client was built against; absent = 1 (legacy clients)."),
    clientVersion: z.string().min(1).max(100).optional().describe("e.g. `web/0.1.0`."),
  }),
);

export const ClientPingEventSchema = named(
  "ClientPingEvent",
  "Keep-alive.",
  z.strictObject({ type: z.literal("ping") }),
);

export const SurfaceSubscribeEventSchema = named(
  "SurfaceSubscribeEvent",
  "Start receiving `surface.frame` events for a thread's surface.",
  z.strictObject({
    type: z.literal("surface.subscribe"),
    threadId: IdSchema,
    surface: SurfaceKindSchema,
  }),
);

export const SurfaceUnsubscribeEventSchema = named(
  "SurfaceUnsubscribeEvent",
  "Stop receiving frames for a thread's surface.",
  z.strictObject({
    type: z.literal("surface.unsubscribe"),
    threadId: IdSchema,
    surface: SurfaceKindSchema,
  }),
);

export const ThreadReadEventSchema = named(
  "ThreadReadEvent",
  "The user has seen a thread (clears its unread count).",
  z.strictObject({ type: z.literal("thread.read"), threadId: IdSchema }),
);

export const EditorActivityEventSchema = named(
  "EditorActivityEvent",
  "Where the user is typing, so the orchestrator never jumps on a half-written task.",
  z.strictObject({
    type: z.literal("editor.activity"),
    notePath: RequestPathSchema,
    line: CountSchema,
  }),
);

export const ClientEventSchema = named(
  "ClientEvent",
  "Every client → server WebSocket message, discriminated by `type`.",
  z.discriminatedUnion("type", [
    ClientHelloEventSchema,
    ClientPingEventSchema,
    SurfaceSubscribeEventSchema,
    SurfaceUnsubscribeEventSchema,
    ThreadReadEventSchema,
    EditorActivityEventSchema,
  ]),
);

export const SERVER_EVENT_TYPES = ServerEventSchema.options.map(
  (option) => option.shape.type.value,
);
export const CLIENT_EVENT_TYPES = ClientEventSchema.options.map(
  (option) => option.shape.type.value,
);
